File: BOJ/JS/done/24479.js
/**
 * https://www.acmicpc.net/problem/24479
 */
const fs = require('fs');
const input = fs.readFileSync('./input.txt').toString().trim().split('\n');

const [n, m, r] = input[0].split(' ').map(Number);
const edges = input.slice(1);

const visited = Array(n).fill(0);
const graph = [];
for (let i = 0; i < n; i++) {
    graph.push([]);
}
edges.forEach((edge) => {
    const [v, u] = edge.split(' ').map(Number);
    graph[v - 1].push(u - 1);
    graph[u - 1].push(v - 1);
});

graph.forEach((row) => {
    row.sort((a, b) => a - b);
});

let count = 1;
function DFS(vertex) {
    if (visited[vertex] != 0) return;
    visited[vertex] = count;
    count++;
    graph[vertex].forEach((next) => {
        DFS(next);
    });
}

DFS(r - 1);
console.log(visited.join('\n'));

File: BOJ/C++/done/24479.cpp
/**
 * https://www.acmicpc.net/problem/24479
 *
 */
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;
vector<int> graph[100000];
vector<int> visited(100000, 0);
int cnt = 1;
void DFS(int vertex)
{
    if (visited[vertex])
        return;
    visited[vertex] = cnt;
    cnt++;
    for (int i = 0; i < graph[vertex].size(); i++)
        DFS(graph[vertex][i]);
}

int main()
{
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);
    int n, m, r, u, v;
    cin >> n >> m >> r;
    for (int i = 0; i < m; i++)
    {
        cin >> u >> v;
        graph[u - 1].push_back(v - 1);
        graph[v - 1].push_back(u - 1);
    }
    for (int i = 0; i < n; i++)
        sort(graph[i].begin(), graph[i].end());

    DFS(r - 1);
    for (int i = 0; i < n; i++)
        cout << visited[i] << '\n';
    return 0;
}